db/filters/base.py

```diff
@@ -0,0 +1,302 @@
+(ReliesOnComparability), as well as its identifier and its human-friendly name.
+"""
+
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Any, List, Union, Type
+from sqlalchemy_filters.exceptions import BadFilterFormat as SABadFilterFormat
+
+
+class PredicateSuperType(Enum):
+    """Every Predicate is either a leaf node in the predicate tree, or a branch node.
+    A leaf node (e.g. EMPTY, EQUAL) has no predicate children, while a branch node
+    (e.g. AND, OR, NOT) always has predicate children."""
+    LEAF = "leaf"
+    BRANCH = "branch"
+
+
+class BranchPredicateType(Enum):
+    NOT = "not"
+    OR = "or"
+    AND = "and"
+
+
+class LeafPredicateType(Enum):
+    """Note that negation is achieved via BranchPredicateType.NOT"""
+    EQUAL = "equal"
+    NOT_EQUAL = "not_equal"
+    GREATER = "greater"
+    GREATER_OR_EQUAL = "greater_or_equal"
+    LESSER = "lesser"
+    LESSER_OR_EQUAL = "lesser_or_equal"
+    EMPTY = "empty"
+    NOT_EMPTY = "not_empty"
+    IN = "in"
+    NOT_IN = "not_in"
+
+
+class ParameterCount(Enum):
+    """Predicates (currently only leaf predicates) can take single parameters (e.g. EQUAL
+    predicate takes a value to check equality against), lists of parameters (e.g. the IN
+    predicate takes a list of values to check membership against), or no paramaters (e.g.
+    the EMPTY predicate).""" 
+    SINGLE = "single"
+    MULTI = "multi"
+    NONE = "none"
+
+
+predicate_types_to_SA_ids = {
+    BranchPredicateType.NOT: 'not',
+    BranchPredicateType.AND: 'and',
+    BranchPredicateType.OR: 'or',
+    LeafPredicateType.EQUAL: 'eq',
+    LeafPredicateType.NOT_EQUAL: 'ne',
+    LeafPredicateType.GREATER: 'gt',
+    LeafPredicateType.GREATER_OR_EQUAL: 'ge',
+    LeafPredicateType.LESSER: 'lt',
+    LeafPredicateType.LESSER_OR_EQUAL: 'le',
+    LeafPredicateType.EMPTY: 'is_null',
+    LeafPredicateType.NOT_EMPTY: 'is_not_null',
+    LeafPredicateType.IN: 'in',
+    LeafPredicateType.NOT_IN: 'not_in',
+}
+
+
+def get_SA_id_from_predicate_type(type: Union[LeafPredicateType, BranchPredicateType]) -> str:
+    if type in predicate_types_to_SA_ids:
+        return predicate_types_to_SA_ids[type]
+    else:
+        raise Exception("This should never happen.")
+
+
+# frozen=True provides immutability
+def frozen_dataclass(f):
+    return dataclass(frozen=True)(f)
+
+
+def static(value):
+    """
+    Declares a static field on a dataclass.
+    """
+    return field(init=False, default=value)
+
+
+@frozen_dataclass
+class Predicate:
+    super_type: PredicateSuperType
+    type: Union[LeafPredicateType, BranchPredicateType]
+    name: str
+    parameter_count: ParameterCount
+
+    def saId(self) -> str:
+        return get_SA_id_from_predicate_type(self.type)
+
+    def __post_init__(self):
+        assert_predicate_correct(self)
+
+
+@frozen_dataclass
+class Leaf(Predicate):
+    super_type: PredicateSuperType = static(PredicateSuperType.LEAF)
```

Comment:
Can we rename `super_type` and `type` here to match the API? `name` and `position`

Reply:
Sure. It's actually `identifier` and `position`. I'd prefer to make the change in the latest filtering PR. Does that work?

Reply:
Sure, that's fine. It shouldn't affect the API, right?